azure-iot-hub/azure/iot/hub/iothub_amqp_client.py

```diff
@@ -100,3 +74,57 @@ def send_message_to_device(self, device_id, message, app_props):
         results = self.amqp_client.send_all_messages(close_on_done=False)
         if uamqp.constants.MessageState.SendFailed in results:
             raise Exception("C2D message send failure")
+
+
+class IoTHubAmqpClientSharedAccessKeyAuth(IoTHubAmqpClientBase):
+    def _generate_auth_token(self, uri, sas_name, sas_value):
+        sas = base64.b64decode(sas_value)
+        expiry = str(int(time.time() + default_sas_expiry))
+        string_to_sign = (uri + "\n" + expiry).encode("utf-8")
+        signed_hmac_sha256 = hmac.HMAC(sas, string_to_sign, hashlib.sha256)
+        signature = urllib.parse.quote(base64.b64encode(signed_hmac_sha256.digest()))
+        return "SharedAccessSignature sr={}&sig={}&se={}&skn={}".format(
+            uri, signature, expiry, sas_name
+        )
+
+    def _build_amqp_endpoint(
+        self,
+        hostname,
+        shared_access_key_name=None,
+        shared_access_key=None,
+    ):
+        hub_name = hostname.split(".")[0]
+        endpoint = "{}@sas.root.{}".format(shared_access_key_name, hub_name)
+        endpoint = quote_plus(endpoint)
+        sas_token = self._generate_auth_token(
+            hostname, shared_access_key_name, shared_access_key + "="
+        )
+        endpoint = endpoint + ":{}@{}".format(quote_plus(sas_token), hostname)
+        return endpoint
+
+    def __init__(self, hostname, shared_access_key_name, shared_access_key):
+        self.endpoint = self._build_amqp_endpoint(
+            hostname, shared_access_key_name, shared_access_key
+        )
+        operation = "/messages/devicebound"
+        target = "amqps://" + self.endpoint + operation
+        self.amqp_client = uamqp.SendClient(target)
+
+
+class IoTHubAmqpClientTokenAuth(IoTHubAmqpClientBase):
+    def __init__(
+        self, hostname, token_credential, token_scope="https://iothubs.azure.net/.default"
+    ):
+        def get_token():
+            result = token_credential.get_token(token_scope)
+            return AccessToken("Bearer " + result.token, result.expires_on)
+
+        auth = uamqp.authentication.JWTTokenAuth(
+            audience=token_scope,
+            uri="https://" + hostname,
+            get_token=get_token,
+            token_type=b"bearer",
+        )
+        auth.update_token()
+        target = "amqps://" + hostname + "/messages/devicebound"
+        self.amqp_client = uamqp.SendClient(target=target, auth=auth)
```

Comment:
Just so we are clear, I doesn't look like to me that this is doing any auto renewal here.  Let me know if I am wrong.  I will let others perform the actual approval, but it lgtm.

Reply:
uamqp should be doing the refresh for us.

Reply:
How is that possible?  The user is passing in a token with an expiration time.  How is uamqp renewing it?

Reply:
We pass in the `get_token` function that uamqp can call. My understanding from looking at the uamqp source code is that `get_token` will be called by uamqp when a renewal is needed.

Reply:
Adam Ling, who works on uamqp, says that the token is refreshed for us.